File: apps/agents/scripts/check_graphql_operations.py
"""Validate agents GraphQL operations against backend root schema fields."""

from __future__ import annotations

import re
from pathlib import Path

REPO_ROOT = Path(__file__).resolve().parents[3]
SCHEMA_PATH = REPO_ROOT / "apps" / "graphql" / "schema.graphql"
OPERATIONS_PATH = REPO_ROOT / "apps" / "agents" / "agents" / "graphql_operations.py"


def parse_root_fields(schema_text: str, root_name: str) -> set[str]:
    lines = schema_text.splitlines()
    start_idx = -1
    for idx, line in enumerate(lines):
        if line.strip() == f"type {root_name} {{":
            start_idx = idx
            break
    if start_idx < 0:
        return set()
    fields: set[str] = set()
    for line in lines[start_idx + 1 :]:
        if line.strip() == "}":
            break
        line = line.strip()
        if not line or line.startswith("#"):
            continue
        field_match = re.match(r"([A-Za-z_][A-Za-z0-9_]*)\s*(\(|:)", line)
        if field_match:
            fields.add(field_match.group(1))
    return fields


def extract_operation_docs(text: str) -> list[str]:
    docs: list[str] = []
    for match in re.finditer(r'"""(.*?)"""', text, re.DOTALL):
        value = match.group(1)
        if re.search(r"\b(query|mutation)\b", value):
            docs.append(value)
    return docs


def operation_type(doc: str) -> str | None:
    if re.search(r"\bmutation\b", doc):
        return "Mutation"
    if re.search(r"\bquery\b", doc):
        return "Query"
    return None


def first_root_field(doc: str) -> str | None:
    if "{" not in doc:
        return None
    body = doc.split("{", 1)[1]
    for line in body.splitlines():
        line = line.strip()
        if not line or line.startswith("#") or line.startswith("}"):
            continue
        line = re.sub(r"^[A-Za-z_][A-Za-z0-9_]*\s*:\s*", "", line)
        match = re.match(r"([A-Za-z_][A-Za-z0-9_]*)", line)
        if match:
            return match.group(1)
    return None


def main() -> None:
    schema_text = SCHEMA_PATH.read_text(encoding="utf-8")
    ops_text = OPERATIONS_PATH.read_text(encoding="utf-8")
    query_fields = parse_root_fields(schema_text, "Query")
    mutation_fields = parse_root_fields(schema_text, "Mutation")
    errors: list[str] = []

    for doc in extract_operation_docs(ops_text):
        op_type = operation_type(doc)
        field = first_root_field(doc)
        if not op_type or not field:
            continue
        root_fields = query_fields if op_type == "Query" else mutation_fields
        if field not in root_fields:
            errors.append(f"Missing {op_type}.{field} for operation:\n{doc.strip().splitlines()[0]}")

    if errors:
        for err in errors:
            print(err)
        raise SystemExit(1)

    print("GraphQL operation validation passed for apps/agents.")


if __name__ == "__main__":
    main()

File: apps/web/scripts/check-graphql-operations.mjs
import fs from 'node:fs/promises'
import path from 'node:path'

const repoRoot = path.resolve(process.cwd(), '..', '..')
const schemaPath = path.join(repoRoot, 'apps', 'graphql', 'schema.graphql')

const operationFiles = [
  path.join(repoRoot, 'apps', 'web', 'lib', 'graphql', 'queries', 'workspace.ts'),
  path.join(repoRoot, 'apps', 'web', 'app', 'api', 'analytics', 'create', 'route.ts'),
  path.join(
    repoRoot,
    'apps',
    'web',
    'app',
    'api',
    'analytics',
    '[analyticsId]',
    'cogs',
    'route.ts',
  ),
]

function parseRootFields(schemaText, rootTypeName) {
  const lines = schemaText.split('\n')
  const start = lines.findIndex((line) => line.trim() === `type ${rootTypeName} {`)
  if (start < 0) return new Set()
  const bodyLines = []
  for (let i = start + 1; i < lines.length; i += 1) {
    const line = lines[i]
    if (line.trim() === '}') break
    bodyLines.push(line)
  }
  const fields = bodyLines
    .map((line) => line.trim())
    .filter(Boolean)
    .filter((line) => !line.startsWith('#'))
    .map((line) => line.match(/^([A-Za-z_][A-Za-z0-9_]*)\s*(\(|:)/)?.[1])
    .filter(Boolean)
  return new Set(fields)
}

function extractGraphqlStrings(fileText) {
  const docs = []
  const regex = /`([\s\S]*?)`/g
  let match
  while ((match = regex.exec(fileText)) !== null) {
    const value = match[1]
    if (/\b(query|mutation)\b/.test(value)) {
      docs.push(value)
    }
  }
  return docs
}

function operationType(doc) {
  if (/\bmutation\b/.test(doc)) return 'Mutation'
  if (/\bquery\b/.test(doc)) return 'Query'
  return null
}

function firstRootField(doc) {
  const body = doc.slice(doc.indexOf('{') + 1)
  const lines = body.split('\n')
  for (const line of lines) {
    const trimmed = line.trim()
    if (!trimmed || trimmed.startsWith('#') || trimmed.startsWith('}')) continue
    const cleaned = trimmed.replace(/^[A-Za-z_][A-Za-z0-9_]*\s*:\s*/, '')
    const field = cleaned.match(/^([A-Za-z_][A-Za-z0-9_]*)/)?.[1]
    if (field) return field
  }
  return null
}

async function main() {
  const schemaText = await fs.readFile(schemaPath, 'utf8')
  const queryFields = parseRootFields(schemaText, 'Query')
  const mutationFields = parseRootFields(schemaText, 'Mutation')
  const errors = []

  for (const opFile of operationFiles) {
    const content = await fs.readFile(opFile, 'utf8')
    const docs = extractGraphqlStrings(content)
    for (const doc of docs) {
      const type = operationType(doc)
      const field = firstRootField(doc)
      if (!type || !field) continue
      const fieldSet = type === 'Query' ? queryFields : mutationFields
      if (!fieldSet.has(field)) {
        errors.push(`${path.relative(repoRoot, opFile)} references missing ${type}.${field}`)
      }
    }
  }

  if (errors.length > 0) {
    console.error('GraphQL operation drift detected:')
    for (const err of errors) console.error(`- ${err}`)
    process.exit(1)
  }
  console.log('GraphQL operation validation passed for apps/web.')
}

await main()
